x-pack/legacy/plugins/epm/server/lib/kibana/index_pattern/install.ts

```diff
@@ -62,71 +63,99 @@ export interface IndexPatternField {
   script?: string;
   lang?: string;
 }
-interface KibanaIndexPattern {
-  [key: string]: string;
-}
-enum IndexPatternType {
+export enum IndexPatternType {
   logs = 'logs',
   metrics = 'metrics',
 }
 
 export async function installIndexPatterns(
-  pkgkey: string,
-  savedObjectsClient: SavedObjectsClientContract
+  savedObjectsClient: SavedObjectsClientContract,
+  pkgkey?: string
 ) {
-  const registryPackageInfo = await Registry.fetchInfo(pkgkey);
-  const registryDatasets = registryPackageInfo.datasets;
-  if (!registryDatasets) return;
+  // get all user installed packages
+  const installedPackages = await getPackageKeysByStatus(
+    savedObjectsClient,
+    InstallationStatus.installed
+  );
+  // add this package
+  if (pkgkey) installedPackages.push(pkgkey);
 
-  const indexPatternTypes = [IndexPatternType.logs, IndexPatternType.metrics];
+  // get each package's registry info
+  const installedPackagesFetchInfoPromise = installedPackages.map(pkg => Registry.fetchInfo(pkg));
+  const installedPackagesInfo = await Promise.all(installedPackagesFetchInfoPromise);
 
+  // for each index pattern type, create an index pattern
+  const indexPatternTypes = [IndexPatternType.logs, IndexPatternType.metrics];
   indexPatternTypes.forEach(async indexPatternType => {
-    const datasets = registryDatasets.filter(dataset => dataset.type === indexPatternType);
-    await createIndexPattern({
-      datasetType: indexPatternType,
-      datasets,
-      registryPackageInfo,
-      savedObjectsClient,
+    // if this is an update because a package is being unisntalled (no pkgkey argument passed) and no other packages are installed, remove the index pattern
+    if (!pkgkey && installedPackages.length === 0) {
+      try {
+        await savedObjectsClient.delete(
+          SAVED_OBJECT_TYPE_INDEX_PATTERN,
+          `epm-ip-${indexPatternType}`
+        );
+      } catch (err) {
+        // index pattern was probably deleted by the user already
+      }
+      return;
+    }
+
+    // get all dataset fields from all installed packages
+    const fields = await getAllDatasetFieldsByType(installedPackagesInfo, indexPatternType);
+
+    const kibanaIndexPattern = createIndexPattern(indexPatternType, fields);
+    // create or overwrite the index pattern
+    await savedObjectsClient.create(SAVED_OBJECT_TYPE_INDEX_PATTERN, kibanaIndexPattern, {
+      id: `epm-ip-${indexPatternType}`,
```

Comment:
I chose what I think would be a unique ID so I could easily target it for deletion and updating